components/widgets/header/header.tsx

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {memo, ReactNode, Fragment, HTMLAttributes} from 'react';
+import classNames from 'classnames';
+
+import './header.scss';
+
+type Props = {
+    heading: ReactNode;
+    level?: 0 | 1 | 2 | 3 | 4 | 5 | 6;
+    subtitle?: ReactNode;
+    right?: ReactNode;
+};
+
+const Headings: Array<keyof JSX.IntrinsicElements | typeof Fragment> = [Fragment, 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'];
+
+const Header = ({
+    level = 0,
+    heading,
+    subtitle,
+    right,
+    ...attrs
+}: Props & HTMLAttributes<HTMLElement>) => {
+    const H = Headings[level];
+    return (
+        <header
+            {...attrs}
+            className={classNames('Header', attrs.className)}
+        >
+            <div className='left'>
+                <H>{heading}</H>
+                {subtitle ? <p>{subtitle}</p> : null}
+            </div>
+            <div className='spacer'/>
+            {right}
+        </header>
+    );
+};
+
+export default memo(Header);
```

Comment:
Do we need to memoize this?

Reply:
Yes, I believe it should, given the number of props and amount of content that may be in them.

It's also my understanding that a `memo`-wrapped functional component is generally equivalent to a `PureComponent` (less the state diffing), and we do usually recommend using that over `Component`.

Reply:
After another look—you're right, we don't need this. The render props make it pointless.